csrc/multidevice/utils.cpp

```diff
@@ -138,39 +129,99 @@ int64_t numDeviceDims(const TensorView* tv) {
 
 bool haveDifferentShardings(
     const TensorView* producer,
-    const TensorView* consumer) {
+    const TensorView* consumer,
+    const IdModel& id_model) {
   // cpu scalars are not required to have a mesh
   if (producer->isCpuScalar() || consumer->isCpuScalar()) {
     return false;
   }
+
   // exit early in the unsharded case for performance
   if (!producer->hasDeviceMesh() && !consumer->hasDeviceMesh()) {
     return false;
   }
+
   // If device mesh are different, the Expr is resharding
-  if (!(producer->getDeviceMesh() == consumer->getDeviceMesh())) {
+  if (producer->getDeviceMesh() != consumer->getDeviceMesh()) {
     return true;
   }
-  // Create a map between producer's and consumer's IterDomains. We iterate
-  // over producer's iterdomain and compare sharding type with consumer's
-  // iterdomain
+
   const std::unordered_map<IterDomain*, IterDomain*>& p2c =
       PairwiseLogicalDomainMap(producer, consumer).mapProducerToConsumer();
-  for (auto p_id : producer->getLogicalDomain()) {
+  std::unordered_set<Val*> mapped_c_root_ids;
+  for (IterDomain* p_id : producer->getLogicalDomain()) {
     const auto i = p2c.find(p_id);
     if (i == p2c.end()) {
-      // This happens e.g. when `p_id` is squeezed or is a product of a
-      // reduction. Even if `p_id` is parallelized on DID, the dimension is
-      // size-1 and doesn't trigger resharding.
+      // This happens e.g. when `p_id` is squeezed. Even if `p_id` is
+      // parallelized on DID, the squeezed dimension is size-1 and doesn't
+      // trigger resharding.
+      continue;
+    }
+    mapped_c_root_ids.insert(i->second);
+  }
+
+  std::unordered_map<ParallelType, IterDomain*> p_parallel_type_to_id;
+  for (IterDomain* p_loop_id : producer->getLoopDomain()) {
+    const ParallelType parallel_type = p_loop_id->getParallelType();
+    if (!isParallelTypeDeviceDim(parallel_type)) {
+      continue;
+    }
+
+    const auto dependencies = IterVisitor::getInputsTo(
+        {p_loop_id},
+        {producer->getLogicalDomain().begin(),
+         producer->getLogicalDomain().end()});
+    if (std::any_of(
+            dependencies.begin(), dependencies.end(), [&](Val* dependency) {
+              return p2c.count(dependency->as<IterDomain>()) > 0;
```

Comment:
I wonder if we would want to do a little more deeper checks. For example, suppose we have:

```
t0 = [i0, b1];

t1 = squeeze(t0, {1}); // t1: [i0]

t0->merge(0, 1);
t0->axis(0)->parallelize(DIDx);
```

The DID-parallelized loop ID of `t0` would be ignored as it uses the squeezed logical ID. However, this case would need resharding, and also if `t1->axis(0)->parallelize(DIDx)`, it would not.

Reply:
The tests you suggested do pass: https://github.com/NVIDIA/Fuser/pull/3421/commits/d32dd6bba1a9fe13cadef4f11d83e63b4a81a42c. This is because t0's i0 is a dependency of the merge and is mapped by PairwiseLogicalDomainMap.

Reply:
Oh, sorry, I misread the code. For some reason, I thought this skips if any of dependencies is missing in the producer-to-consumer map.